crates/egui/src/viewport.rs

```diff
@@ -0,0 +1,493 @@
+        self.0.short_debug_format().fmt(f)
+    }
+}
+
+impl ViewportId {
+    /// The `ViewportId` of the root viewport.
+    pub const ROOT: Self = Self(Id::null());
+
+    pub fn from_hash_of(source: impl std::hash::Hash) -> Self {
+        Self(Id::new(source))
+    }
+}
+
+impl From<ViewportId> for Id {
+    #[inline]
+    fn from(id: ViewportId) -> Self {
+        id.0
+    }
+}
+
+impl nohash_hasher::IsEnabled for ViewportId {}
+
+/// A fast hash map from [`ViewportId`] to `T`.
+pub type ViewportMap<T> = nohash_hasher::IntMap<ViewportId, T>;
+
+// ----------------------------------------------------------------------------
+
+/// This will deref to [`Self::this`].
+#[derive(Debug, Hash, Clone, Copy, PartialEq, Eq)]
+pub struct ViewportIdPair {
+    pub this: ViewportId,
+    pub parent: ViewportId,
+}
+
+impl Default for ViewportIdPair {
+    fn default() -> Self {
+        Self {
+            this: ViewportId::ROOT,
+            parent: ViewportId::ROOT,
+        }
+    }
+}
+
+impl ViewportIdPair {
+    /// The `ViewportIdPair` of the root viewport, which is its own parent.
+    pub const ROOT: Self = Self {
+        this: ViewportId::ROOT,
+        parent: ViewportId::ROOT,
+    };
+}
+
+/// The user-code that shows the ui in the viewport, used for "async" viewports.
+pub type ViewportUiCallback = dyn Fn(&Context) + Sync + Send;
+
+/// Render the given viewport, calling the given ui callback.
+pub type ViewportRenderSyncCallback =
+    dyn for<'a> Fn(&Context, ViewportBuilder, ViewportIdPair, Box<dyn FnOnce(&Context) + 'a>);
+
+/// Control the building of a new egui viewport (i.e. native window).
+///
+/// The fields are public, but you should use the builder pattern to set them,
+/// and that's where you'll find the documentation too.
+///
+/// Since egui is immediate mode, `ViewportBuilder` is accumulative in nature.
+/// Setting any option to `None` means "keep the current value",
+/// or "Use the default" if it is the first call.
+#[derive(PartialEq, Eq, Clone)]
+#[allow(clippy::option_option)]
+pub struct ViewportBuilder {
+    pub id: ViewportId,
+
+    /// The title of the vieweport.
+    /// `eframe` will use this as the title of the native window.
+    pub title: Option<String>,
+
+    /// This is wayland only. See [`Self::with_name`].
+    pub name: Option<(String, String)>,
+
+    pub position: Option<Option<Pos2>>,
+    pub inner_size: Option<Option<Vec2>>,
+    pub fullscreen: Option<bool>,
+    pub maximized: Option<bool>,
+    pub resizable: Option<bool>,
+    pub transparent: Option<bool>,
+    pub decorations: Option<bool>,
+    pub icon: Option<Option<Arc<ColorImage>>>,
+    pub active: Option<bool>,
+    pub visible: Option<bool>,
+    pub title_hidden: Option<bool>,
+    pub titlebar_transparent: Option<bool>,
+    pub fullsize_content_view: Option<bool>,
+    pub min_inner_size: Option<Option<Vec2>>,
+    pub max_inner_size: Option<Option<Vec2>>,
+    pub drag_and_drop: Option<bool>,
```

Comment:
Why are some of these options of options?

Reply:
That is because the `ViewportBuilder` is like a patch that is applied over the existing ViewportBuilder!
The `egui_winit::changes_between_builders` will make ViewportCommands and say if the window needs recreation!

So for `min_inner_size`
If first Option is None that means that will remain the same!
If first Option is Some and then second one is None that means that can have any size!
If first Option is Some and then second one is Some(10) that means that the Viewport will have the min size of 10

Why:
This was implemented when i was experimenting with making any popup a native window, but if forgot for what was used!

But if was not like that, only the first ViewportBuilder was applied and others only if the viewport was destroyed by the os!

Now that i rethink this: i will say that the `egui_winit::changes_between_builders` needs to be moved in egui because do not uses the winit!